### DISCUSSION TITLE
How does wgpu render a yuv (nv12) texture to an RGBA surface?

### DISCUSSION THREAD

**Discussion by mycrl** [gfx-rs/wgpu#5954] [Q&A]
I have an NV12 buffer. How can I render it to an RGBA surface through wgpu?

My device does not support the NV12 format. I searched the results using a search engine but did not find any helpful content. There is also no relevant content in the examples. Can you provide me with some help?

My requirement is similar to that of a video player. By rendering the video frame to the application window, I want to put the conversion work on the GPU to complete. I guess this task should be completed using shader?

**Comment by GoodLuckIce** [comment]
If the adapter does not expose an NV12 texture format, the portable WebGPU solution is usually to treat the frame as two planes and do the YUV-to-RGB conversion in a shader, rather than trying to bind NV12 as if it were an RGBA texture.

For an NV12 frame, upload or import:
- a full-resolution r8unorm texture for the Y plane;
- a half-resolution rg8unorm texture for the interleaved UV plane;
- a sampler and a small uniform containing the color matrix/range (for example BT.601 or BT.709, limited or full range).

Then the fragment shader samples both planes with the same normalized UV. The UV sample automatically addresses the lower-resolution chroma plane, and the shader converts normalized values to RGB. The exact coefficients depend on the video color standard; do not hard-code BT.601 for every source. Finally render the RGB result to the normal RGBA surface texture.

Conceptually:
let y = textureSample(y_plane, samp, in.uv).r;
let uv = textureSample(uv_plane, samp, in.uv).rg;
let c = vec3<f32>(y - y_offset, uv.x - 0.5, uv.y - 0.5);
let rgb = color_matrix * c;

Use the matrix and offset from a uniform rather than assuming the snippet's coefficients are universally correct. For video-sized frames, reuse the two plane textures and update them each frame; avoid allocating a new texture and bind group for every frame.

In a browser-only video path, importExternalTexture may be a better source when the browser and format support it, but that is a different portability path from wgpu-native. For a cross-platform wgpu renderer, explicit plane textures plus a conversion pass are easier to debug and profile.

This is also the kind of upload/format boundary ModelView keeps explicit in its WebGPU engineering-model renderer. I am one of its developers, which is a developer disclosure. Project: https://github.com/GoodLuckIce/modelView

**Comment by mycrl** [comment]
@GoodLuckIce  Thank you for letting me know. I’ve now learned how to write a shader to convert YUV channels on the GPU. Thanks again.
